src/napari/layers/_layer_actions.py

```diff
@@ -56,7 +56,7 @@ def _convert(ll: LayerList, type_: str) -> None:
     for lay in list(ll.selection):
         idx = ll.index(lay)
         if isinstance(lay, Shapes) and type_ == 'labels':
-            data = lay.to_labels()
+            data = lay.to_labels(labels_shape=ll.extent.world[-1] + 1)
```

Comment:
```suggestion
            data = lay.to_labels(labels_shape=lay.world_to_data(ll.extent.world[-1]) + 1)
```
I thought about this a bit more and I think transforming from world back to data for setting the **shape** of the labels layer is the correct way to handle this.
I tested with a translation and scale -- not both at the same time -- and it worked as expected!
Would be good to sanity check a bit more, it's a bit late 😴 

Also, it took me a while to get the `+1` -- I was very confused.
Then I realized that extent goes from 0 to `shape - 1` while this kwarg is **shape** so we need the `+1` my suggestion is to add a comment about that, unless folks feel it's too obvious and trivial and i'm just 😴